src/v3/plugins/github/graphView.test.js

```diff
@@ -0,0 +1,383 @@
+        });
+        it("src must be author in edge address", () => {
+          const otherAuthor = {type: "USERLIKE", login: "wchargin"};
+          const authorsEdge = GE.createEdge.authors(otherAuthor, issue);
+          (authorsEdge: any).src = GN.toRaw(userlike);
+          const g = exampleWithParents().addEdge(authorsEdge);
+          expect(() => new GraphView(g)).toThrow("Invariant: Expected src");
+        });
+        it("dst must be content in edge address", () => {
+          const authorsEdge = GE.createEdge.authors(userlike, issue);
+          (authorsEdge: any).dst = GN.toRaw(pull);
+          const g = exampleWithParents().addEdge(authorsEdge);
+          expect(() => new GraphView(g)).toThrow("Invariant: Expected dst");
+        });
+      });
+      describe("merged as edges", () => {
+        const commitAddress = NodeAddress.fromParts([
+          "sourcecred",
+          "git",
+          "COMMIT",
+          "hash",
+        ]);
+        it("src must be a pull", () => {
+          // $ExpectFlowError
+          const badEdge = GE.createEdge.mergedAs(issue, commitAddress);
+          failsForEdge(badEdge);
+        });
+        it("dst must be commit address", () => {
+          const badEdge = GE.createEdge.mergedAs(pull, GN.toRaw(repo));
+          failsForEdge(badEdge);
+        });
+        it("src must be pull in edge address", () => {
+          const otherPull = {type: "PULL", repo, number: "143"};
+          const mergedAs = GE.createEdge.mergedAs(otherPull, commitAddress);
+          (mergedAs: any).src = GN.toRaw(pull);
+          const g = exampleWithParents()
+            .addNode(commitAddress)
+            .addEdge(mergedAs);
+          expect(() => new GraphView(g)).toThrow("Invariant: Expected src");
+        });
+      });
+      describe("references edges", () => {
+        it("src must be a TextContentAddress", () => {
+          // $ExpectFlowError
+          const badEdge = GE.createEdge.references(userlike, pull);
+          failsForEdge(badEdge);
+        });
+        it("src must be the referrer in edge address", () => {
+          const references = GE.createEdge.references(issue, review);
+          (references: any).src = GN.toRaw(pull);
+          const g = exampleWithParents().addEdge(references);
+          expect(() => new GraphView(g)).toThrow("Invariant: Expected src");
+        });
+        it("dst must be referent in edge address", () => {
+          const references = GE.createEdge.references(issue, review);
+          (references: any).dst = GN.toRaw(pull);
+          const g = exampleWithParents().addEdge(references);
+          expect(() => new GraphView(g)).toThrow("Invariant: Expected dst");
+        });
+      });
+      describe("has parent edges", () => {
+        it("must satisfy specific hom relationships", () => {
+          const g = new Graph()
+            .addNode(GN.toRaw(repo))
+            // $ExpectFlowError
+            .addEdge(GE.createEdge.hasParent(repo, repo));
```

Comment:
A repository can’t have a parent at all. Can we instead provide a valid
child and a valid parent, like `hasParent(pullComment, issue)`? (In
other words, this test currently checks “general” hom relationships, not
“specific” ones.)

Reply:
That is harder than it sounds, because such cases are already checked by the node-level parent invariant. Suppose that I try the following example:

```js
          const g = new Graph()
            .addNode(GN.toRaw(repo))
            .addNode(GN.toRaw(issue))
            .addEdge(GE.createEdge.hasParent(issue, repo))
            .addNode(GN.toRaw(pullComment))
            .addEdge(GE.createEdge.hasParent(pullComment, issue));
          expect(() => new GraphView(g)).toThrow("Invariant: Edge");
```

This triggers the node-wise invariant checker to throw, because `pullComment` _is_ a pull comment by virtue of having a pull as its address-encoded parent, and the `hasParent` edge is a mismatch on its address-encoded parent.

I can get away with using `repo,repo` because the node-wise invariant checker doesn't expect repo to have a parent at all.

It's technically possible to surgically construct addresses that are valid except with mutilated prefixes, but the effort-to-power ratio is unfavorable. Making no change.

Reply:
Noted, and reasonable. Thank you for the explanation.